dev/src/lobster/flat_gltf.h

```diff
@@ -0,0 +1,789 @@
+  glPixelStorei(GL_UNPACK_ALIGNMENT,
+       pitch % 8 == 0 ? 8 // most efficient
+    : (pitch % 4 == 0 ? 4 // common value
+    : (pitch % 2 == 0 ? 2 // dubious efficiency
+                      : 1 // least efficient
+  )));
+
+  // assume an sRGBA image, which OpenGL automatically linearizes before passing
+  // to every shader or load said sRGBA image as linear RGBA so fragments are
+  // output 1:1 without lighting effects
+  // https://www.khronos.org/registry/OpenGL-Refpages/es3.0/html/glTexImage2D.xhtml
+  // https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glTexImage2D.xhtml
+  // http://opengl.datenwolf.net/gltut/html/Texturing/Tutorial%2016.html
+  // TODO: Lobster's framebuffer is currently not gamma-corrected :/
+  // Lobster currently loads textures as GL_RGBA8, which may or may not be physically correct
+  glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, // GL_SRGB8_ALPHA8
+    width, height, 0, pixfmts[channels - 1], GL_UNSIGNED_BYTE, data);
+
+  if (mipmap)
+    glGenerateMipmap(GL_TEXTURE_2D);
+
+  tex.success = true;
+  tex.width = width;
+  tex.height = height;
+  tex.channels = channels;
+  tex.mipmap = mipmap;
+
+  return tex;
+}
+
+void SetTexture(Texture tex, GLuint texSlot) {
+  if (texSlot == 0)
+    drawTexture0 = tex;
+  if (tex.success) {
+    glActiveTexture(GL_TEXTURE0 + texSlot);
+    glBindTexture(GL_TEXTURE_2D, tex.texture);
+  }
+}
+
+} // namespace fgl
+
+namespace fgltf {
+
+bool verbose = true;
+
+#ifndef _VERBOUT_
+#define _VERBOUT_                                                              \
+  if (verbose)                                                                 \
+  std::cout
+#endif
+
+#ifndef _ERROUT_
+#define _ERROUT_ std::cerr
+#endif
+
+#ifndef MAX_SCENE_NODES
+#define MAX_SCENE_NODES 1000000
+#endif
+
+typedef struct {
+  int baseColorImage;
+  GLfloat metallicFactor;
+  GLfloat roughnessFactor;
+} SceneMaterial;
+
+SceneMaterial newSceneMaterial = {-1, 0.0f, 0.5f};
+
+typedef struct {
+  fgl::Mesh mesh;
+  int material;
+  int mode;
+} ScenePrim;
+
+typedef std::vector<ScenePrim> SceneMesh;
+
+typedef struct {
+  int mesh;
+  float3 translation;
+  float4 rotation;
+  float3 scale;
+  std::vector<int> children;
+} SceneNode;
+
+typedef struct {
+  int node;
+  int type;
+  int interpolation;
+  std::vector<float3> translations;
+  std::vector<float4> rotations;
+  std::vector<float3> scales;
+  std::vector<float> weights;
+} SceneAnimationData;
+
+typedef struct {
+  std::string name;
+  std::vector<SceneAnimationData> channels;
+} SceneAnimation;
+
+typedef struct {
+  std::vector<SceneMaterial> materials;
+  std::vector<SceneMesh> meshes;
+  std::vector<SceneNode> nodes;
+  std::vector<SceneAnimation> animations;
+  std::vector<int> roots;
+} Scene;
+
+Scene newScene = {{}, {}, {}, {}, {}};
+
+typedef struct {
+  std::string path;
+  fgl::Texture texture;
+} Image;
+
+std::vector<Image> images;
+
+int LoadImage(std::string gltfPath, std::string imgName, bool filter = true) {
```

Comment:
Can this be replaced with `CreateTextureFromFile` ?

Reply:
Nope. It caches all the textures in video memory with associate filename strings so that all subsequent glTF loads recycle textures that are already loaded. It's the coolest way to deal with texture loading. 😎

Reply:
You can do the same with the result from `CreateTextureFromFile`.

And Lobster does this kind of caching in user code: https://github.com/aardappel/lobster/blob/master/lobster/modules/texture.lobster

Reply:
Well, it looks like I can just pop `CreateTextureFromFile` right into where I'm using the STB stuff and it should work. :)